chore: refactor MergeSettings to not be used in the MegaCircuitBuilder

maramihali opened the pull request:
Refactor how the MergeSettings are used in the EccOpQueue so they have a smaller code footprint and can be removed from the builder. This is achieved by accumulating ops in the current subtable separately and doing the actual merge at the beginning of merge proving.

Commit:
[empty] Start merge-train. Choo choo.

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

federicobarbacovi committed:
feat!: Merge with degree check (#15562)

We modify the Merge protocol so that it enforces that the subtable
polynomial `t_j` has degree smaller than `subtable_size` as read from
the proof.

Closes https://github.com/AztecProtocol/barretenberg/issues/1442

**Details**

As per the linked issue, we want to support ecc operations in app
circuits. To ensure that app ecc ops do not modify ecc ops that happened
before them, we need to ensure that the subtable length `t.size()` is
smaller than the constant `l` by which we right shift `T_prev`. This is
to ensure that `t + X^{l-1} T_prev` is indeed the polynomial
corresponding to the column `t || T_prev`.

We enforce this degree check in the merge protocol by requiring the
prover to commit to `g(X) := X^{l-1} t(1/X)` and provide openings `c`,
`d` of `t`, `g` at challenges `1/kappa`, `kappa`, respectively, for
which we check `c * kappa^{l-1} = d`.

To save on the number of MSMs performed, we use Shplonk to verify the
following claims:
- `t(X)` opens to `c` at `1/kappa`
- `p(X) = t(X) + X^{l-1} T_prev(X) - T(X)` opens to `0` at `kappa`
- `g(X) := X^{l-1} t(1/X)` opens to `d` at `kappa`

---------

Co-authored-by: AztecBot <tech@aztecprotocol.com>
Co-authored-by: sergei iakovenko <105737703+iakovenkos@users.noreply.github.com>
Co-authored-by: ludamad <adam.domurad@gmail.com>
Co-authored-by: Suyash Bagad <suyash@aztecprotocol.com>
Co-authored-by: Jonathan Hao <jonathan@aztec-labs.com>
Co-authored-by: ledwards2225 <98505400+ledwards2225@users.noreply.github.com>
Co-authored-by: Raju Krishnamoorthy <krishnamoorthy@gmail.com>
Co-authored-by: notnotraju <raju@aztec-labs.com>
Co-authored-by: Lucas Xia <lucasxia01@gmail.com>
Co-authored-by: Khashayar Barooti <khashayar@aztecprotocol.com>
Co-authored-by: Jean M <132435771+jeanmon@users.noreply.github.com>
Co-authored-by: Alex Gherghisan <alexghr@users.noreply.github.com>
Co-authored-by: Santiago Palladino <spalladino@users.noreply.github.com>
Co-authored-by: Santiago Palladino <santiago@aztec-labs.com>
Co-authored-by: ludamad <domuradical@gmail.com>
Co-authored-by: maramihali <mara@aztecprotocol.com>

Sarkoxed committed:
fix: Fix the docker build action for fuzzing (#15719)

johnathan79717 committed:
fix: restore accidentally deleted files (#15724)

Restore accidentally deleted files in a previous PR.
https://github.com/AztecProtocol/aztec-packages/pull/15562

iakovenkos committed:
fix: civc wasm memory regression (#15722)

In my [previous
PR](https://github.com/AztecProtocol/aztec-packages/pull/15427), I moved
the `GateSeparatorPolynomial` into the SumcheckProver class as a member,
which extended its lifetime and increased memory usage in CIVC by the
size of this polynomial, e.g. by 21 MB in
`ecdsar1+transfer_0_recursions+sponsored_fpc` bench

johnathan79717 committed:
feat: mmap backed polynomials (#15531)

This makes `SharedShiftedVirtualZeroesArray` holds an abstract
BackingMemory. When `BB_SLOW_LOW_MEMORY=1` or when the
`--slow_low_memory` runtime flag is enabled, `BackingMemory` will be
`FileBackedMemory`. The memory is `mmap`ed to physical files so in
memory-constrained environment, it won't be OOM. Otherwise,
`BackingMemory` is `AlignedMemory` and the behavior is unchanged.

See https://github.com/AztecProtocol/barretenberg/issues/1456 for more
contexts.

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

ludamad committed:
feat(bbapi): CLI uses bbapi CIVC (#15702)

- Move the CIVC cli to use the new bbapi API
- add placeholder for UH implementation
- rename bbrpc => bbapi
- bundle some CI changes, auto rebase script and bench_ivc tweaks

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
[empty] Start merge-train. Choo choo.

ledwards2225 committed:
chore!: Remove circuit_size from all VKs (#15747)

- Removes `circuit_size` from all VKs (only store `log_circuit_size`)
- Updates `compute_public_input_delta()` to correctly take witness
values and compute `dyadic_size` via `pow()`

Addresses most of
https://github.com/AztecProtocol/barretenberg/issues/1283 but one
instance remains

federicobarbacovi committed:
chore: Introduce data structures to hold inputs/outputs of the Merge verification (#15735)

Introduce the classes `SubtableWitnessCommitments` which stores
`t_commitments` (and in the future `T_prev_commitments`) and
`WitnessCommitments`, which stores `t_commitments`, `T_commitments` (and
in the future `T_prev_commitments`). The role of these classes is to
facilitate the introduction of consistency checks between two
consecutive merges, see
[#1351](https://github.com/AztecProtocol/barretenberg/issues/1351).

- The value of `t_commitments` is populated by extracting the
commitments from the VK of the PG verifier.
- (Future work) The value of `T_prev_commitments` is populated by taking
the previous value of `T_commitments`
- The value of `T_commitments` is populated by the Merge verifier

---------

Co-authored-by: AztecBot <tech@aztecprotocol.com>
Co-authored-by: sergei iakovenko <105737703+iakovenkos@users.noreply.github.com>
Co-authored-by: ludamad <adam.domurad@gmail.com>
Co-authored-by: Suyash Bagad <suyash@aztecprotocol.com>
Co-authored-by: Jonathan Hao <jonathan@aztec-labs.com>
Co-authored-by: ledwards2225 <98505400+ledwards2225@users.noreply.github.com>
Co-authored-by: Raju Krishnamoorthy <krishnamoorthy@gmail.com>
Co-authored-by: notnotraju <raju@aztec-labs.com>
Co-authored-by: Lucas Xia <lucasxia01@gmail.com>
Co-authored-by: Khashayar Barooti <khashayar@aztecprotocol.com>
Co-authored-by: Jean M <132435771+jeanmon@users.noreply.github.com>
Co-authored-by: Alex Gherghisan <alexghr@users.noreply.github.com>
Co-authored-by: Santiago Palladino <spalladino@users.noreply.github.com>
Co-authored-by: Santiago Palladino <santiago@aztec-labs.com>
Co-authored-by: ludamad <domuradical@gmail.com>
Co-authored-by: maramihali <mara@aztecprotocol.com>

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

johnathan79717 committed:
chore: replace asserts with runtime errors. (#15671)

1. Asserts now throw runtime errors in non-debug builds.
1. `BB_ASSERT_*` now throw_or_abort in non-debug builds.
2. Replace `ASSERT` in tests with `ASSERT_TRUE` or other gtest
functions.
3. Replace `EXPECT_DEATH` and `ASSERT_DEATH` in tests with
`EXPECT_THROW_OR_ABORT` or `ASSERT_THROW_OR_ABORT`.
4. Remove `#ifdef NDEBUG` around the above in tests since they work in
non-debug builds now.
5. Asserts in `constexpr`s uses info instead of sstream.

Fixes https://github.com/AztecProtocol/barretenberg/issues/1460

Commit:
chore: kernels start with eq and reset (#15734)

As a measure of precaution in case of misconfigured ecc ops in an app
circuit that might attempt, for example, to link ecc ops between apps
and kernels, the subtable of ecc ops in each kernel circuit will start
with an eq and reset op. This cannot happen when the API for creating
ecc ops is used properly as the logic in `goblin_element` ensures an eq
and reset op is present at the end of each `batch_mul`, function also
used for `operator+` and `operator-`. Moreover, the
`add_gates_to_ensure_nonzero_polynomials` makes sure that an eq and
reset exists if no other ecc ops are encountered in the circuit.

---------

Co-authored-by: AztecBot <tech@aztecprotocol.com>
Co-authored-by: federicobarbacovi <171914500+federicobarbacovi@users.noreply.github.com>
Co-authored-by: sergei iakovenko <105737703+iakovenkos@users.noreply.github.com>
Co-authored-by: ludamad <adam.domurad@gmail.com>
Co-authored-by: Suyash Bagad <suyash@aztecprotocol.com>
Co-authored-by: Jonathan Hao <jonathan@aztec-labs.com>
Co-authored-by: ledwards2225 <98505400+ledwards2225@users.noreply.github.com>
Co-authored-by: Raju Krishnamoorthy <krishnamoorthy@gmail.com>
Co-authored-by: notnotraju <raju@aztec-labs.com>
Co-authored-by: Lucas Xia <lucasxia01@gmail.com>
Co-authored-by: Khashayar Barooti <khashayar@aztecprotocol.com>
Co-authored-by: Jean M <132435771+jeanmon@users.noreply.github.com>
Co-authored-by: Alex Gherghisan <alexghr@users.noreply.github.com>
Co-authored-by: Santiago Palladino <spalladino@users.noreply.github.com>
Co-authored-by: Santiago Palladino <santiago@aztec-labs.com>
Co-authored-by: ludamad <domuradical@gmail.com>
Co-authored-by: Sarkoxed <75146596+Sarkoxed@users.noreply.github.com>

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
redesign settings

Commit:
Merge remote-tracking branch 'origin/merge-train/barretenberg' into mm/builder-without-settings

Commit:
remove modificadions to sol

Commit:
cleanup

Commit:
fix tests

Commit:
remove info

Commit:
[empty] Start merge-train. Choo choo.

johnathan79717 committed:
chore: flag to update inputs when precomputed vk fails (#15828)

This is the first part to fix
https://github.com/AztecProtocol/barretenberg/issues/1464. Once we have
this, we can add an option to
`test_civc_standalone_vks_havent_changed.sh` that generates inputs with
this option.

ledwards2225 reviewed (changes requested):
Much better! Thanks for doing this. Left some suggestions for how to avoid the cost of physically prepending the tables

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
[empty] Start merge-train. Choo choo.

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

ledwards2225 committed:
chore: remove public_inputs from DeciderVK (#15832)

Removes explicit storage of `public_inputs` in the DeciderVK. The public
inputs are folded as part of the wire polynomials, not as individual
elements so its not meaningful for the verifier accumulator to store
them explicitly (which has recently led to some confusion in IVC state
mocking). They are now stored directly in the oink/PG verifiers where
they are populated from the proof and used to reconstruct the IO data.

johnathan79717 committed:
chore: delete LegacyPolynomial (#15847)

`LegacyPolynomial` is only used in `polynomial.test.cpp`, where we can
use `Polynomial` instead. Some test cases were deleted because there was
no corresponding method to test in `Polynomial`:
- `factor_roots`: this tests multiple roots but `Polynomial` only has a
method for a single root.
- `RightShift`: There's no similar method for `Polynomial`.
`right_shifted` works differently and points to a different block of
memory of the same size.
- `evaluate_mles`: it compares results between `Polynomial` and
`LegacyPolynomial` so makes no sense when `LegacyPolynomial` is deleted.

Commit:
Merge remote-tracking branch 'origin/merge-train/barretenberg' into mm/builder-without-settings

Commit:
Merge remote-tracking branch 'origin/merge-train/barretenberg' into mm/builder-without-settings

Commit:
use std::dequeue

Commit:
fix broken build caused by merge

ledwards2225 reviewed (approved):
Nice, thanks!

Commit:
Merge remote-tracking branch 'origin/merge-train/barretenberg' into mm/builder-without-settings
